File: working_hours/application/correction_service.py
"""Correction write use cases."""
from __future__ import annotations

import datetime
import uuid

from ..infrastructure.data import (
    _append_correction,
    add_pending,
    list_pending,
    remove_pending,
)


def add_correction(emp_id: str, name: str, dept: str, timestamp: str) -> None:
    _append_correction(f"ADD\t{emp_id}\t{name}\t{dept}\t{timestamp}\t1")


def delete_correction(emp_id: str, name: str, dept: str, timestamp: str) -> None:
    _append_correction(f"DEL\t{emp_id}\t{name}\t{dept}\t{timestamp}\t1")


def edit_correction(
    emp_id: str, name: str, dept: str, old_ts: str, new_ts: str
) -> None:
    _append_correction(f"EDIT\t{emp_id}\t{name}\t{dept}\t{old_ts}\t{new_ts}\t1")


def bulk_delete(items: list[dict]) -> None:
    for item in items:
        delete_correction(item["emp_id"], item["name"], item["dept"], item["timestamp"])


def queue_correction(
    action: str,
    emp_id: str,
    name: str,
    dept: str,
    timestamp: str,
    new_timestamp: str | None,
    submitted_by: str,
) -> None:
    item = {
        "id": str(uuid.uuid4()),
        "action": action,
        "emp_id": emp_id,
        "name": name,
        "dept": dept,
        "timestamp": timestamp,
        "new_timestamp": new_timestamp,
        "submitted_at": datetime.datetime.now().isoformat(timespec="seconds"),
        "submitted_by": submitted_by,
    }
    add_pending(item)


def get_pending() -> list[dict]:
    return list_pending()


def approve_pending(item_id: str) -> bool:
    item = remove_pending(item_id)
    if item is None:
        return False
    if item["action"] == "ADD":
        _append_correction(
            f"ADD\t{item['emp_id']}\t{item['name']}\t{item['dept']}\t{item['timestamp']}\t1"
        )
    elif item["action"] == "EDIT":
        _append_correction(
            f"EDIT\t{item['emp_id']}\t{item['name']}\t{item['dept']}"
            f"\t{item['timestamp']}\t{item['new_timestamp']}\t1"
        )
    return True


def reject_pending(item_id: str) -> bool:
    return remove_pending(item_id) is not None

File: packages/server/src/application/correctionService.ts
import { randomUUID } from 'crypto'
import { addPending, addToHistory, loadHistory, loadPending, removePending, removeHistoryItem } from '../infrastructure/data.js'
import type { HistoryItem, PendingItem } from '../infrastructure/data.js'

function recordHistory(action: 'ADD' | 'EDIT' | 'DEL', empId: string, name: string, dept: string, timestamp: string, newTimestamp: string | null, appliedBy: string): void {
  addToHistory({
    id: randomUUID(),
    action,
    emp_id: empId,
    name,
    dept,
    timestamp,
    new_timestamp: newTimestamp,
    applied_at: new Date().toISOString().slice(0, 19),
    applied_by: appliedBy,
  })
}

export function addCorrection(empId: string, name: string, dept: string, timestamp: string, appliedBy: string): void {
  recordHistory('ADD', empId, name, dept, timestamp, null, appliedBy)
}

export function deleteCorrection(empId: string, name: string, dept: string, timestamp: string, appliedBy: string): void {
  recordHistory('DEL', empId, name, dept, timestamp, null, appliedBy)
}

export function editCorrection(empId: string, name: string, dept: string, oldTs: string, newTs: string, appliedBy: string): void {
  recordHistory('EDIT', empId, name, dept, oldTs, newTs, appliedBy)
}

export function bulkDelete(items: { emp_id: string; name: string; dept: string; timestamp: string }[], appliedBy: string): void {
  for (const item of items) deleteCorrection(item.emp_id, item.name, item.dept, item.timestamp, appliedBy)
}

export function queueCorrection(
  action: 'ADD' | 'EDIT' | 'DEL',
  empId: string,
  name: string,
  dept: string,
  timestamp: string,
  newTimestamp: string | null,
  submittedBy: string
): void {
  const item: PendingItem = {
    id: randomUUID(),
    action,
    emp_id: empId,
    name,
    dept,
    timestamp,
    new_timestamp: newTimestamp,
    submitted_at: new Date().toISOString().slice(0, 19),
    submitted_by: submittedBy,
  }
  addPending(item)
}

export function getPending(): PendingItem[] {
  return loadPending()
}

export function approvePending(itemId: string, approvedBy: string): boolean {
  const item = removePending(itemId)
  if (!item) return false
  recordHistory(item.action, item.emp_id, item.name, item.dept, item.timestamp, item.new_timestamp, approvedBy)
  return true
}

export function rejectPending(itemId: string): boolean {
  return removePending(itemId) !== null
}

export function getHistory(): HistoryItem[] {
  return loadHistory().slice().reverse()
}

export function revertCorrection(itemId: string): boolean {
  return removeHistoryItem(itemId)
}

export function undoCorrection(itemId: string): boolean {
  return removeHistoryItem(itemId)
}
